controller/api/destination/endpoint_translator.go

```diff
@@ -287,6 +303,120 @@ func (et *endpointTranslator) sendClientAdd(set watcher.AddressSet) {
+// address is removed. When the pod address is removed, closeEndpointPolicy
+// will close the client.
+//
+// The policy server client handles Recv errors and signals the updates
+// routine to shutdown.
+func (et *endpointTranslator) watchEndpointPolicy(set watcher.AddressSet) {
+	for _, addr := range set.Addresses {
+		// If the address is not backed by a pod then there will be no policy
+		// server to watch.
+		if addr.Pod == nil {
+			continue
+		}
+
+		opaquePorts, ok, err := getPodOpaquePortsAnnotations(addr.Pod)
+		if err != nil {
+			et.log.Errorf("failed to get opaque ports annotation for pod: %s", err)
+		}
+
+		// If the opaque ports annotation was not set, then use the
+		// endpoint translator's default value.
+		if !ok {
+			opaquePorts = et.defaultOpaquePorts
+		}
+
+		skippedInboundPorts, err := getPodSkippedInboundPortsAnnotations(addr.Pod)
+		if err != nil {
+			et.log.Errorf("failed to get ignored inbound ports annotation for pod: %s", err)
+		}
+
+		// Create a client which watches for changes to the inbound policy for
+		// portSpec.
+		policyPortSpec := &policyPb.PortSpec{
+			Workload: fmt.Sprintf("%s:%s", addr.Pod.Namespace, addr.Pod.Name),
+			Port:     addr.Port,
+		}
+		client, err := et.policyClient.WatchPort(et.stream.Context(), policyPortSpec)
+		if err != nil {
+			et.log.Errorf("failed to create policy server watch for %s:%d: %s", policyPortSpec.Workload, policyPortSpec.Port, err)
+		}
+		et.log.Infof("Establishing watch on policy server %s:%d", policyPortSpec.Workload, policyPortSpec.Port)
+
+		// policyPb.PortSpec cannot be used as a map key, so we create an
+		// equivalent port spec with comparable keys.
+		portSpec := &portSpec{
+			workload: policyPortSpec.Workload,
+			port:     policyPortSpec.Port,
+		}
+		et.policyWatches[*portSpec] = client
+
+		// Receive policy server updates for portSpec.
+		//
+		// Running this in the background allows it to handle Recv errors when
+		// the stream context is canceled. When canceled, this routine will
+		// complete.
+		updates := make(chan *policyPb.Server)
+		done := make(chan struct{})
+		go func() {
+			for {
+				update, err := client.Recv()
+				if err != nil {
+					et.log.Debugf("Shutting down policy server updates for %s:%d", portSpec.workload, portSpec.port)
+					close(done)
+					break
+				}
+				updates <- update
+			}
+		}()
+
+		// Wait for stream cancellation or policy server updates.
+		//
+		// When an update is available, create a new weighted addr that may
+		// include an opaque transport port as specified by a policy server.
+		addr := addr
+		go func() {
+			for {
+				select {
+				case <-done:
+					et.log.Infof("Stopping policy server watch for %s:%d", portSpec.workload, portSpec.port)
+					return
+				case <-et.stream.Context().Done():
+					et.log.Infof("Stopping policy server watch for %s:%d", portSpec.workload, portSpec.port)
+					return
+				case update := <-updates:
```

Comment:
instead of having a separate done channel, can you use the 2 value receive here to check if the updates channel is close? ` case update, ok := <- updates:`  and then close the updates channel when client.Recv encounters an error.